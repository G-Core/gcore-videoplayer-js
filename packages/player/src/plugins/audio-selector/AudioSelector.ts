import { Events, UICorePlugin, template } from '@clappr/core'
import { AudioTrack } from '@clappr/core/types/base/playback/playback.js'
import assert from 'assert'

import { CLAPPR_VERSION } from '../../build.js'

import pluginHtml from '../../../assets/audio-selector/track-selector.ejs'
import '../../../assets/audio-selector/style.scss'
import audioArrow from '../../../assets/icons/old/quality-arrow.svg'
import { ZeptoResult } from '../../types.js'
import { MediaControl } from '../media-control/MediaControl.js'

const VERSION: string = '2.22.4'

// const T = 'plugins.audiotracks'

/**
 * `PLUGIN` that makes possible to switch audio tracks via the media control UI.
 * @beta
 * @remarks
 * The plugin is activated when there are multiple audio tracks available.
 * The plugin adds a button showing the current audio track and a dropdown to switch to another audio track.
 * Depends on:
 *
 * - {@link MediaControl}
 */
export class AudioTracks extends UICorePlugin {
  private currentTrack: AudioTrack | null = null

  private tracks: AudioTrack[] = []

  /**
   * @internal
   */
  get name() {
    return 'audio_selector' // TODO rename to audiotracks
  }

  /**
   * @internal
   */
  get supportedVersion() {
    return { min: CLAPPR_VERSION }
  }

  /**
   * @internal
   */
  static get version() {
    return VERSION
  }

  private static readonly template = template(pluginHtml)

  /**
   * @internal
   */
  override get attributes() {
    return {
      class: 'media-control-audiotracks',
    }
  }

  /**
   * @internal
   */
  override get events() {
    return {
      'click [data-audiotracks-select]': 'onTrackSelect',
      'click #audiotracks-button': 'toggleContextMenu',
    }
  }

  /**
   * @internal
   */
  override bindEvents() {
    this.listenToOnce(this.core, Events.CORE_READY, this.onCoreReady)
    this.listenTo(
      this.core,
      Events.CORE_ACTIVE_CONTAINER_CHANGED,
      this.onActiveContainerChanged,
    )
  }

  private onCoreReady() {
    const mediaControl = this.core.getPlugin('media_control')
    assert(mediaControl, 'media_control plugin is required')
    this.listenTo(mediaControl, Events.MEDIACONTROL_RENDERED, () => {
      mediaControl.putElement('audiotracks', this.$el)
    })
    this.listenTo(mediaControl, Events.MEDIACONTROL_HIDE, this.hideMenu)
  }

  private onActiveContainerChanged() {
    this.currentTrack = null
    this.listenTo(
      this.core.activeContainer,
      Events.CONTAINER_AUDIO_AVAILABLE,
      (tracks: AudioTrack[]) => {
        this.currentTrack =
          tracks.find((track) => track.kind === 'main') ?? null // TODO test
        this.tracks = tracks
        this.render()
      },
    )
    this.listenTo(
      this.core.activeContainer,
      Events.CONTAINER_AUDIO_CHANGED,
      (track: AudioTrack) => {
        this.currentTrack = track
        this.highlightCurrentTrack()
        this.buttonElement().removeClass('changing')
        this.updateText()
      },
    )
  }

  private shouldRender() {
    // Render is called from the parent class constructor so tracks aren't available
    // Only care if we have at least 2 to choose from
    return this.tracks?.length > 1
  }

  /**
   * @internal
   */
  override render() {
    if (!this.shouldRender()) {
      return this
    }

    const mediaControl = this.core.getPlugin('media_control') as MediaControl
    this.$el.html(
      AudioTracks.template({
        tracks: this.tracks,
        title: this.getTitle(),
        icon: audioArrow,
      }),
    )
    this.updateText()
    this.highlightCurrentTrack()

    return this
  }

  private onTrackSelect(event: MouseEvent) {
    const id = (event.target as HTMLElement)?.dataset?.audiotracksSelect
    if (id) {
      this.selectAudioTrack(id)
    }
    this.hideMenu()
    event.stopPropagation()
    return false
  }

  private selectAudioTrack(id: string) {
    this.startTrackSwitch()
    this.core.activeContainer.switchAudioTrack(id)
    this.updateText()
  }

  private hideMenu() {
    this.$el.find('#audiotracks-select').addClass('hidden')
  }

  private toggleContextMenu() {
    this.$el.find('#audiotracks-select').toggleClass('hidden')
  }

  private buttonElement(): ZeptoResult {
    return this.$('button')
  }

  private buttonElementText(): ZeptoResult {
    return this.$('button .audio-text')
  }

  private trackElement(id?: string): ZeptoResult {
    return (
      this.$(
        '#audiotracks-select a' +
          (id !== undefined ? `[data-audiotracks-select="${id}"]` : ''),
      ) as ZeptoResult
    ).parent()
  }

  private getTitle(): string {
    if (!this.currentTrack) {
      return ''
    }
    return this.currentTrack.label || this.currentTrack.language
  }

  private startTrackSwitch() {
    this.buttonElement().addClass('changing')
  }

  private updateText() {
    if (!this.currentTrack) {
      return
    }
    this.buttonElementText().text(this.currentTrack.label)
  }

  private highlightCurrentTrack() {
    this.trackElement().removeClass('current')
    this.trackElement().find('a').removeClass('gcore-skin-active')

    if (this.currentTrack) {
      this.trackElement(this.currentTrack.id)
        .addClass('current')
        .find('a')
        .addClass('gcore-skin-active')
    }
  }
}
