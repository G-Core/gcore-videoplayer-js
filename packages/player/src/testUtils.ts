import { $, UICorePlugin } from '@clappr/core'
import Events from 'eventemitter3'
import { vi } from 'vitest'

export function createMockCore(
  options: Record<string, unknown> = {},
  container: any = createMockContainer(options),
) {
  const el = document.createElement('div')
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    $el: $(el),
    activePlayback: container.playback,
    activeContainer: container,
    i18n: {
      t: vi.fn().mockImplementation((key: string) => key),
    },
    options: {
      ...options,
    },
    configure: vi.fn(),
    getPlaybackType: vi.fn(),
    getPlugin: vi.fn(),
    load: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockPlugin() {
  return Object.assign(new Events(), {
    enable: vi.fn(),
    disable: vi.fn(),
  })
}

export function createSpinnerPlugin() {
  return Object.assign(createMockPlugin(), {
    show: vi.fn(),
    hide: vi.fn(),
  })
}

export function createMockPlayback(name = 'mock') {
  const emitter = new Events()
  return Object.assign(emitter, {
    name,
    currentLevel: -1,
    el: document.createElement('video'),
    dvrEnabled: false,
    dvrInUse: false,
    isAudioOnly: false,
    levels: [],
    consent: vi.fn(),
    play: vi.fn(),
    pause: vi.fn(),
    stop: vi.fn(),
    destroy: vi.fn(),
    seek: vi.fn(),
    seekPercentage: vi.fn(),
    getDuration: vi.fn().mockImplementation(() => 100),
    enterPiP: vi.fn(),
    exitPiP: vi.fn(),
    getPlaybackType: vi.fn(),
    getStartTimeOffset: vi.fn().mockImplementation(() => 0),
    getCurrentTime: vi.fn().mockImplementation(() => 0),
    isHighDefinitionInUse: vi.fn().mockImplementation(() => false),
    mute: vi.fn(),
    unmute: vi.fn(),
    volume: vi.fn(),
    configure: vi.fn(),
    attemptAutoPlay: vi.fn().mockImplementation(() => true),
    canAutoPlay: vi.fn().mockImplementation(() => true),
    onResize: vi.fn().mockImplementation(() => true),
    setPlaybackRate: vi.fn(),
    switchAudioTrack: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockContainer(
  options: Record<string, unknown> = {},
  playback: any = createMockPlayback(),
) {
  const el = playback.el
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    playback,
    options: {
      ...options,
    },
    $el: $(el),
    disableMediaControl: vi.fn(),
    enableMediaControl: vi.fn(),
    enterPiP: vi.fn(),
    exitPiP: vi.fn(),
    getDuration: vi.fn().mockReturnValue(0),
    getPlugin: vi.fn(),
    getPlaybackType: vi.fn(),
    getStartTimeOffset: vi.fn().mockReturnValue(0),
    isDvrInUse: vi.fn().mockReturnValue(false),
    isDvrEnabled: vi.fn().mockReturnValue(false),
    isHighDefinitionInUse: vi.fn().mockReturnValue(false),
    isPlaying: vi.fn().mockReturnValue(false),
    play: vi.fn(),
    seek: vi.fn(),
    seekPercentage: vi.fn(),
    switchAudioTrack: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockMediaControl(core: any) {
  const mediaControl = new UICorePlugin(core)
  // TODO <div class="media-control-layer">
  mediaControl.$el.html(
    `<div class="media-control-left-panel" data-media-control></div>
    <div class="media-control-right-panel" data-media-control></div>
    <div class="media-control-center-panel" data-media-control></div>`,
  )
  // @ts-ignore
  mediaControl.mount = vi.fn()
  // @ts-ignore
  mediaControl.container = core.activeContainer
  // @ts-ignore
  mediaControl.toggleElement = vi.fn()
  vi.spyOn(mediaControl, 'trigger')
  core.$el.append(mediaControl.$el)
  return mediaControl
}

export function createMockBottomGear(core: any) {
  const plugin: any = new UICorePlugin(core)
  plugin.getItem = vi.fn()
  plugin.addItem = vi.fn().mockImplementation((name: string, $el: any) => {
    const existing = plugin.$el.find(`[data-${name}]`)
    if (existing.length) {
      return existing
    }
    return $('<li></li>')
      .attr(`data-${name}`, '')
      .append($el)
      .appendTo(plugin.$el)
  })
  plugin.refresh = vi.fn()
  return plugin
}
