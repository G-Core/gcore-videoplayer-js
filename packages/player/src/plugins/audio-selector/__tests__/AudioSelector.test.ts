import { beforeEach, describe, expect, it, vi } from 'vitest'
import { Events } from '@clappr/core'

import { AudioTracks } from '../AudioSelector'

import { createMockCore, createMockMediaControl } from '../../../testUtils'
// import { LogTracer, Logger, setTracer } from '@gcorevideo/utils'

// Logger.enable('*')
// setTracer(new LogTracer('AudioSelector.test'))

const TRACKS = [
  { id: '1', label: 'English', language: 'en', track: {} },
  { id: '2', label: 'Spanish', language: 'es', track: {} },
]

describe('AudioSelector', () => {
  let core: any
  let mediaControl: any
  let audioSelector: AudioTracks
  beforeEach(() => {
    core = createMockCore()
    mediaControl = createMockMediaControl(core)
    core.getPlugin = vi.fn().mockImplementation((name: string) => {
      if (name === 'media_control') return mediaControl
      return null
    })
    audioSelector = new AudioTracks(core)
    core.emit(Events.CORE_READY)
    core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED, core.activeContainer)
  })
  describe('before media control is rendererd', () => {
    beforeEach(() => {
      emitTracksAvailable(core, TRACKS)
    })
    it('should not attach to the media control', () => {
      expect(mediaControl.mount).not.toHaveBeenCalledWith(
        'audiotracks',
        expect.anything(),
      )
    })
  })
  describe('when media control is rendered', () => {
    beforeEach(() => {
      mediaControl.trigger(Events.MEDIACONTROL_RENDERED)
    })
    it('should attach to the media control', () => {
      expect(mediaControl.mount).toHaveBeenCalledWith(
        'audiotracks',
        audioSelector.$el,
      )
    })
  })
  describe('given that audio tracks are available', () => {
    beforeEach(() => {
      emitTracksAvailable(core, TRACKS)
    })
    it('should render button', () => {
      expect(audioSelector.$el.find('#audiotracks-button').length).toBe(1)
    })
    it('should render menu hidden', () => {
      expect(audioSelector.el.innerHTML).toMatchSnapshot()
      expect(
        audioSelector.$el.find('#audiotracks-select').hasClass('hidden'),
      ).toBe(true)
      const trackItems = audioSelector.$el.find('#audiotracks-select li')
      expect(trackItems.length).toBe(2)
      expect(trackItems.eq(0).text().trim()).toBe('English')
      expect(trackItems.eq(1).text().trim()).toBe('Spanish')
    })
    describe('when button is clicked', () => {
      beforeEach(() => {
        audioSelector.$el.find('#audiotracks-button').click()
      })
      it('should show menu', () => {
        expect(audioSelector.$el.html()).toMatchSnapshot()
        expect(
          audioSelector.$el.find('#audiotracks-select').hasClass('hidden'),
        ).toBe(false)
      })
      describe('when audio track is selected', () => {
        beforeEach(() => {
          audioSelector.$el
            .find('#audiotracks-select [data-audiotracks-select="2"]')
            .click()
        })
        it('should switch to the selected audio track', () => {
          expect(core.activeContainer.switchAudioTrack).toHaveBeenCalledWith(
            '2',
          )
        })
        it('should hide the menu', () => {
          expect(audioSelector.$el.html()).toMatchSnapshot()
          expect(
            audioSelector.$el.find('#audiotracks-select').hasClass('hidden'),
          ).toBe(true)
        })
        it('should add changing class to the button', () => {
          expect(
            audioSelector.$el.find('#audiotracks-button').hasClass('changing'),
          ).toBe(true)
        })
        describe('when current audio track changes', () => {
          beforeEach(() => {
            core.activePlayback.currentAudioTrack =
              core.activePlayback.audioTracks[1]
            core.activePlayback.emit(
              Events.PLAYBACK_AUDIO_CHANGED,
              core.activePlayback.currentAudioTrack,
            )
            core.activeContainer.emit(
              Events.CONTAINER_AUDIO_CHANGED,
              core.activePlayback.currentAudioTrack,
            )
          })
          it('should update button class', () => {
            expect(
              audioSelector.$el
                .find('#audiotracks-button')
                .hasClass('changing'),
            ).toBe(false)
          })
          it('should update button label', () => {
            expect(
              audioSelector.$el
                .find('#audiotracks-button')
                .text()
                .replace(/\/assets.*\.svg/g, '')
                .trim(),
            ).toBe('Spanish')
          })
          it('should highlight the selected menu item', () => {
            const selectedItem = audioSelector.$el.find(
              '#audiotracks-select .current',
            )
            expect(selectedItem.text().trim()).toBe('Spanish')
            expect(
              selectedItem
                .find('a[data-audiotracks-select]')
                .hasClass('gcore-skin-active'),
            ).toBe(true)
          })
          it('should unhighlight any previously highlighted menu item', () => {
            expect(
              audioSelector.$el.find('#audiotracks-select li.current').length,
            ).toBe(1)
            expect(
              audioSelector.$el.find(
                '#audiotracks-select a.gcore-skin-active[data-audiotracks-select]',
              ).length,
            ).toBe(1)
          })
        })
      })
    })
  })
  describe('when audio tracks are not available', () => {
    it('should not render the button', () => {
      expect(audioSelector.$el.find('#audiotracks-button').length).toBe(0)
      expect(audioSelector.$el.find('#audiotracks-select').length).toBe(0)
    })
  })
})

function emitTracksAvailable(core: any, tracks: any[]) {
  core.activePlayback.audioTracks = tracks
  core.activePlayback.currentAudioTrack = tracks[0]
  core.activePlayback.emit(
    Events.PLAYBACK_AUDIO_AVAILABLE,
    core.activePlayback.audioTracks,
  )
  core.activeContainer.emit(
    Events.CONTAINER_AUDIO_AVAILABLE,
    core.activePlayback.audioTracks,
  )
}
