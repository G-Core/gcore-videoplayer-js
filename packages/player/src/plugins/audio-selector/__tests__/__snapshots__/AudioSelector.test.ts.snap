// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AudioSelector > given that audio tracks are available > should render menu hidden 1`] = `
"<button data-gplayer-mc-audiotracks-button="" class="gcore-skin-button-color" id="gplayer-mc-audiotracks-button" aria-haspopup="menu" aria-expanded="false">
    <span class="audio-text"></span> <span class="audio-arrow">/assets/icons/old/quality-arrow.svg</span>
</button>
<ul class="gcore-skin-bg-color menu hidden" id="media-control-audiotracks-select" role="menu">
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-media-control-audiotracks-select="1" role="menuitemradio" aria-checked="false">
                English
            </a>
        </li>
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-media-control-audiotracks-select="2" role="menuitemradio" aria-checked="false">
                Spanish
            </a>
        </li>
    
</ul>
"
`;

exports[`AudioSelector > given that audio tracks are available > when button is clicked > should show menu 1`] = `
"<button data-gplayer-mc-audiotracks-button="" class="gcore-skin-button-color" id="gplayer-mc-audiotracks-button" aria-haspopup="menu" aria-expanded="true">
    <span class="audio-text"></span> <span class="audio-arrow">/assets/icons/old/quality-arrow.svg</span>
</button>
<ul class="gcore-skin-bg-color menu" id="media-control-audiotracks-select" role="menu">
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-media-control-audiotracks-select="1" role="menuitemradio" aria-checked="false">
                English
            </a>
        </li>
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-media-control-audiotracks-select="2" role="menuitemradio" aria-checked="false">
                Spanish
            </a>
        </li>
    
</ul>
"
`;

exports[`AudioSelector > given that audio tracks are available > when button is clicked > when audio track is selected > should hide the menu 1`] = `
"<button data-gplayer-mc-audiotracks-button="" class="gcore-skin-button-color changing" id="gplayer-mc-audiotracks-button" aria-haspopup="menu" aria-expanded="true">
    <span class="audio-text"></span> <span class="audio-arrow">/assets/icons/old/quality-arrow.svg</span>
</button>
<ul class="gcore-skin-bg-color menu hidden" id="media-control-audiotracks-select" role="menu">
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-media-control-audiotracks-select="1" role="menuitemradio" aria-checked="false">
                English
            </a>
        </li>
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-media-control-audiotracks-select="2" role="menuitemradio" aria-checked="false">
                Spanish
            </a>
        </li>
    
</ul>
"
`;

exports[`AudioTracks > given that audio tracks are available > should render menu hidden 1`] = `
"<button data-gplayer-mc-audiotracks-button="" class="gcore-skin-button-color" id="gplayer-mc-audiotracks-button" aria-haspopup="menu" aria-expanded="false">
    <span class="audio-text"></span> <span class="audio-arrow">/assets/icons/old/quality-arrow.svg</span>
</button>
<ul class="gcore-skin-bg-color menu hidden" id="media-control-audiotracks-select" role="menu">
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-media-control-audiotracks-select="1" role="menuitemradio" aria-checked="false">
                English
            </a>
        </li>
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-media-control-audiotracks-select="2" role="menuitemradio" aria-checked="false">
                Spanish
            </a>
        </li>
    
</ul>
"
`;

exports[`AudioTracks > given that audio tracks are available > when button is clicked > should show menu 1`] = `
"<button data-gplayer-mc-audiotracks-button="" class="gcore-skin-button-color" id="gplayer-mc-audiotracks-button" aria-haspopup="menu" aria-expanded="true">
    <span class="audio-text"></span> <span class="audio-arrow">/assets/icons/old/quality-arrow.svg</span>
</button>
<ul class="gcore-skin-bg-color menu" id="media-control-audiotracks-select" role="menu">
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-media-control-audiotracks-select="1" role="menuitemradio" aria-checked="false">
                English
            </a>
        </li>
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-media-control-audiotracks-select="2" role="menuitemradio" aria-checked="false">
                Spanish
            </a>
        </li>
    
</ul>
"
`;

exports[`AudioTracks > given that audio tracks are available > when button is clicked > when audio track is selected > should hide the menu 1`] = `
"<button data-gplayer-mc-audiotracks-button="" class="gcore-skin-button-color changing" id="gplayer-mc-audiotracks-button" aria-haspopup="menu" aria-expanded="true">
    <span class="audio-text"></span> <span class="audio-arrow">/assets/icons/old/quality-arrow.svg</span>
</button>
<ul class="gcore-skin-bg-color menu hidden" id="media-control-audiotracks-select" role="menu">
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-media-control-audiotracks-select="1" role="menuitemradio" aria-checked="false">
                English
            </a>
        </li>
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-media-control-audiotracks-select="2" role="menuitemradio" aria-checked="false">
                Spanish
            </a>
        </li>
    
</ul>
"
`;
